spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java

```diff
@@ -201,4 +223,84 @@ static MatcherCondition parse(Object target, String method, Object[] args, boole
       return null;
     }
   }
+
+  private static class CollectionCondition {
+
+    private final String method;
+    private final Object left;
+    private final Object right;
+
+    public CollectionCondition(String method, Object left, Object right) {
+      this.method = method;
+      this.left = left;
+      this.right = right;
+    }
+
+    void verify(ErrorCollector errorCollector, @Nullable List<Object> values, @Nullable String text, int line, int column, @Nullable String message) {
+      String description = null;
+      int idxActual = values.indexOf(left);
+      int idxExpected = values.indexOf(right);
+
+      if (left instanceof Iterable && right instanceof Iterable) {
+        if (SpockRuntime.MATCH_COLLECTIONS_AS_SET.equals(method)) {
+          Set<?> actual = GroovyRuntimeUtil.coerce(left, LinkedHashSet.class);
+          Set<?> expected = GroovyRuntimeUtil.coerce(right, LinkedHashSet.class);
```

Comment:
Can we detect whether `right` was declared as a literal and fail if it had duplicate elements?

Reply:
It's probably unintended but I'm not sure whether we should fail in such cases. WDYT?

Reply:
Fair enough, let's not do it then.